File: 0001-0500/0005/0005.longest-palindromic-substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#

# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if len(s) == 0:
            return 0
        max_len = 1
        start=0
        for i in range(len(s)):
        	if i-max_len >=1 and s[i-max_len-1:i+1]==s[i-max_len-1:i+1][::-1]:
        		start=i-max_len-1
        		max_len+=2
        		continue

        	if i-max_len >=0 and s[i-max_len:i+1]==s[i-max_len:i+1][::-1]:
        		start=i-max_len
        		max_len+=1
        return s[start:start+max_len]
# @lc code=end

File: 0001-0500/0005/0005.longest-palindromic-substring.cs
/*
 * @lc app=leetcode id=5 lang=csharp
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
public class Solution {
    public string LongestPalindrome(string s) {
        int maxLength = 0, startIndex = 0;
        for (int i = 0; i < s.Length; i++) {
            int start = i, end = i;
            while(end < s.Length-1 && s[start] == s[end+1])
                end++;
            
            while(end < s.Length-1 && start > 0 && s[start-1] == s[end+1]) {
                start--; 
                end++;
            }
            if(maxLength < end - start + 1) {
                maxLength = end - start + 1;
                startIndex = start;
            }
        }
        return s.Substring(startIndex, maxLength);
    }
}
// @lc code=end

